Assets/Scripts/Player/PlayerHide.cs

```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.Events;
+
+public class PlayerHide : MonoBehaviour
+{
+    [SerializeField] UnityEvent hidEvent;
+    [SerializeField] UnityEvent unHidEvent;
+    public bool IsHidden { get; private set; }
+    Rigidbody2D _rigidbody;
+    CircleCollider2D _playerCollider;
+
+    void Awake()
+    {
+        _rigidbody = GetComponent<Rigidbody2D>();
+        _playerCollider = GetComponent<CircleCollider2D>();
+    }
+
+    void Update()
+    {
+        var velocity = _rigidbody.velocity;
+        var isMoving = velocity.x > 0.01f || velocity.y > 0.01f;
+        var isPressingButtons = Input.anyKeyDown;
+        if (isMoving || isPressingButtons)
+        {
+            IsHidden = false;
+            unHidEvent.Invoke();
+        }
+        
+        if (Input.GetButtonDown("Hide") && isOnSpot())
+        {
+            IsHidden = true;
+            hidEvent.Invoke();
+        }
+    }
+
+    bool isOnSpot()
+    {
+        var colliders = new List<Collider2D>();
+        _playerCollider.GetContacts(colliders);
+        return colliders.Any(collider => collider.gameObject.CompareTag("HidingSpot"));
```

Comment:
Not that it matters much in our case, but this seems like a rather expensive operation to be calling every Update - it has allocation on line 40 that is done every time the method is called. I propose moving the list out of the method and just clearing and updating the contents of it, so as to avoid this issue.

Reply:
true. We could run most of the cheks when pressing hide key rather then all the time. We can also use onTriggerEnter and Exit for onSpot check